### PR TITLE
Fix scrollbar dragging into overscroll when not allowed

### LINKED ISSUE #85579: Mac desktop scroll physics should be range maintaining
On Mac, scrolling should not over scroll and bounce at the boundaries of scrollable content. We prevent over scrolling with a mouse directly in the handling for pointer scrolls, however, this still leaves open the opportunity to over scroll through other means. 

One example of this is through the scrollbar. While you cannot scroll **_with a mouse scroll wheel_** into over scroll on mac, if you grab the scrollbar, you can. Shown here:

https://user-images.githubusercontent.com/16964204/123874822-58521800-d8fe-11eb-92e9-0050cd13dc7f.mov

The scrollbar checks with the scroll physics to see if dragging into overscroll is allowed here:

https://github.com/flutter/flutter/blob/0825f49ed4fabef3c6dfbea29be3480ea6bd54cd/packages/flutter/lib/src/widgets/scrollbar.dart#L1339

---

So either:
- Bouncing scroll physics should not allow overscrolling on mac
- or when on mac we should not use bouncing scroll physics
--- 

Also of note, when using a trackpad, over scrolling **_is_** allowed natively on mac. Since trackpad support is not official yet (tracked here: https://github.com/flutter/flutter/issues/23604), this may be something to revisit then. Currently we treat trackpad gestures the same as mouse pointer scrolls, so they are not distinguishable right now AFAIK.

Here is the trackpad over scrolling natively on Big Sur:

https://user-images.githubusercontent.com/16964204/123875607-a0256f00-d8ff-11eb-8527-3882aa4a5611.mov

---

### When running this on macOS, so the scrollbar is applied by the ScrollBehavior.

```dart
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(title: Text('Demo')),
        body: ListView.builder(
          itemCount: 50,
          itemBuilder: (BuildContext context, int index) {
            return ListTile(title: Text('Index $index'));
          },
        ),
      )
    );
  }
}
```

---

## `flutter doctor`
```
[✓] Flutter (Channel master, 2.3.0-17.0.pre.556, on macOS 11.3.1 20E241 darwin-x64, locale en)
    • Flutter version 2.3.0-17.0.pre.556 at /Users/katelovett/_github/flutter
    • Upstream repository git@github.com:flutter/flutter.git
    • Framework revision 76d5e62f42 (24 hours ago), 2021-06-28 15:26:04 -0700
    • Engine revision 5d287a2061
    • Dart version 2.14.0 (build 2.14.0-250.0.dev)

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at /Users/katelovett/Library/Android/sdk
    • Platform android-S, build-tools 30.0.3
    • Java binary at: /Applications/Android Studio with Blaze.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6916264)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.4, Build version 12D4e
    • CocoaPods version 1.10.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.2)
    • Android Studio at /Applications/Android Studio with Blaze.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6916264)

[✓] IntelliJ IDEA Ultimate Edition (version 2020.3.3)
    • IntelliJ at /Applications/IntelliJ UE (stable).app
    • Flutter plugin version 57.0.3
    • Dart plugin version 203.8173

[✓] VS Code (version 1.57.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.23.0

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-x64     • macOS 11.3.1 20E241 darwin-x64
    • Chrome (web)    • chrome • web-javascript • Google Chrome 91.0.4472.114

• No issues found!
```

### PR DESCRIPTION
**Pull request by Piinks**
Fixes https://github.com/flutter/flutter/issues/85579

This fixes unexpected behavior from dragging the scrollbar when the ScrollPhysics allow overscrolling. On iOS, you should be able to drag the scrollbar into overscroll, but on other platforms this should not be possible. This clamps the new scroll position when dragging the thumb to stay within the range of the ScrollPosition.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter/lib/src/widgets/scrollbar.dart [hunk 1]
```diff
@@ -1371,7 +1372,23 @@ class RawScrollbarState<T extends RawScrollbar> extends State<T> with TickerProv
...
+
+      // The physics may allow overscroll when actually *scrolling*, but
+      // dragging on the scrollbar does not always allow us to enter overscroll.
+      switch(ScrollConfiguration.of(context).getPlatform(context)) {
+        case TargetPlatform.android:
+        case TargetPlatform.fuchsia:
+        case TargetPlatform.linux:
+        case TargetPlatform.macOS:
+        case TargetPlatform.windows:
+          newPosition = newPosition.clamp(0.0, position.maxScrollExtent);
```
Comment (xu-baolin) [inline packages/flutter/lib/src/widgets/scrollbar.dart:1385]: We can use the 'Bounce physics' on non-iOS platform, why not allow over scrolling in this case?

Comment (xu-baolin) [reply packages/flutter/lib/src/widgets/scrollbar.dart:1385]: Oh, I see，to be consistent with the behavior of the platform.

Comment (xu-baolin) [reply packages/flutter/lib/src/widgets/scrollbar.dart:1385]: How about let drag overscroll on the mobile platform, you know that some developers want to keep the ios behavior on Android phones uses bounce physics.

Comment (Piinks) [reply packages/flutter/lib/src/widgets/scrollbar.dart:1385]: I think that sounds good. I'll update. Thanks! :)

File: packages/flutter/lib/src/widgets/scrollbar.dart [hunk 2]
```diff
@@ -1371,7 +1372,23 @@ class RawScrollbarState<T extends RawScrollbar> extends State<T> with TickerProv
     if (scrollOffsetGlobal != position.pixels) {
       // Ensure we don't drag into overscroll if the physics do not allow it.
       final double physicsAdjustment = position.physics.applyBoundaryConditions(position, scrollOffsetGlobal);
-      position.jumpTo(scrollOffsetGlobal - physicsAdjustment);
+      double newPosition = scrollOffsetGlobal - physicsAdjustment;
+
+      // The physics may allow overscroll when actually *scrolling*, but
+      // dragging on the scrollbar does not always allow us to enter overscroll.
+      switch(ScrollConfiguration.of(context).getPlatform(context)) {
+        case TargetPlatform.android:
```
Comment (Piinks) [inline packages/flutter/lib/src/widgets/scrollbar.dart:1380]:
```suggestion
```

File: packages/flutter/lib/src/widgets/scrollbar.dart [hunk 3]
```diff
@@ -1371,7 +1372,23 @@ class RawScrollbarState<T extends RawScrollbar> extends State<T> with TickerProv
...
+        case TargetPlatform.fuchsia:
+        case TargetPlatform.linux:
+        case TargetPlatform.macOS:
+        case TargetPlatform.windows:
+          newPosition = newPosition.clamp(0.0, position.maxScrollExtent);
+          break;
+        case TargetPlatform.iOS:
```
Comment (Piinks) [inline packages/flutter/lib/src/widgets/scrollbar.dart:1387]:
```suggestion
        case TargetPlatform.iOS:
        case TargetPlatform.android:
```

File: packages/flutter/lib/src/widgets/scrollbar.dart [hunk 4]
```diff
@@ -1371,7 +1372,23 @@ class RawScrollbarState<T extends RawScrollbar> extends State<T> with TickerProv
...
+          // We can only drag the scrollbar into overscroll on iOS.
```
Comment (Piinks) [inline packages/flutter/lib/src/widgets/scrollbar.dart:1388]:
```suggestion
          // We can only drag the scrollbar into overscroll on mobile
          // platforms, and only if the physics allow it.
```

File: packages/flutter/test/widgets/scrollbar_test.dart [hunk 5]
```diff
@@ -990,6 +990,65 @@ void main() {
...
+    expect(
+      find.byType(RawScrollbar),
+      paints
+        ..rect(rect: const Rect.fromLTRB(794.0, 0.0, 800.0, 600.0))
+        ..rect(
+          rect: const Rect.fromLTRB(794.0, 0.0, 800.0, 90.0),
+          color: const Color(0x66BCBCBC),
+        ),
+    );
+  }, variant: TargetPlatformVariant.only(TargetPlatform.macOS));
```
Comment (gspencergoog) [inline packages/flutter/test/widgets/scrollbar_test.dart:1050]: Did you also want to test other platforms to make sure the overscroll _doesn't_ happen there?

**Review by gspencergoog — APPROVED** [review]
32384589-a60f0e74-c078-11e7-9bc1-e5b5287aea9d

**Pull request merged**